scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -675,31 +718,61 @@ void BPF_STRUCT_OPS(layered_running, struct task_struct *p)
 		layer->vtime_now = p->scx.dsq_vtime;
 
 	cctx->current_preempt = layer->preempt;
+	cctx->current_exclusive = layer->exclusive;
 	tctx->started_running_at = bpf_ktime_get_ns();
+
+	/*
+	 * If this CPU is transitioning from running an exclusive task to a
+	 * non-exclusive one, the sibling CPU has likely been idle. Wake it up.
+	 */
+	if (cctx->prev_exclusive && !cctx->current_exclusive) {
+		s32 sib = sibling_cpu(scx_bpf_task_cpu(p));
+		struct cpu_ctx *sib_cctx;
+
+		/*
+		 * %SCX_KICK_IDLE would be great here but we want to support
+		 * older kernels. Let's use racy custom idle flag instead.
+		 */
+		if (sib >= 0 && (sib_cctx = lookup_cpu_ctx(sib)) &&
+		    sib_cctx->maybe_idle) {
+			gstat_inc(GSTAT_EXCL_WAKEUP, cctx);
+			scx_bpf_kick_cpu(sib, 0);
+		}
+	}
+
+	cctx->maybe_idle = false;
 }
 
 void BPF_STRUCT_OPS(layered_stopping, struct task_struct *p, bool runnable)
 {
 	struct cpu_ctx *cctx;
 	struct task_ctx *tctx;
+	struct layer *layer;
+	s32 lidx;
 	u64 used;
-	u32 layer;
 
 	if (!(cctx = lookup_cpu_ctx(-1)) || !(tctx = lookup_task_ctx(p)))
 		return;
 
-	layer = tctx->layer;
-	if (layer >= nr_layers) {
-		scx_bpf_error("invalid layer %u", layer);
+	lidx = tctx->layer;
+	if (!(layer = lookup_layer(lidx)))
 		return;
-	}
 
 	used = bpf_ktime_get_ns() - tctx->started_running_at;
-	cctx->layer_cycles[layer] += used;
+	if (used < layer->min_exec_ns) {
+		lstat_inc(LSTAT_MIN_EXEC, layer, cctx);
+		lstat_add(LSTAT_MIN_EXEC_NS, layer, cctx, layer->min_exec_ns - used);
+		used = layer->min_exec_ns;
+	}
+
+	cctx->layer_cycles[lidx] += used;
 	cctx->current_preempt = false;
+	cctx->prev_exclusive = cctx->current_exclusive;
+	cctx->current_exclusive = false;
 
 	/* scale the execution time by the inverse of the weight and charge */
 	p->scx.dsq_vtime += used * 100 / p->scx.weight;
+	cctx->maybe_idle = true;
```

Comment:
Why is this maybe_idle? Just because it's racy to access from other cores?

Reply:
Yeah, it's not just racy but also inaccurate because it gets asserted between `stopping()` of the prev task and `running()` of the next task. It's a good enough but not great approximation.